File: 509-Fibonacci-Number.py
class Solution:
    def nthfib(self, n, dp):
        if dp[n] != -1:
            return dp[n]
        else:
            dp[n] = self.nthfib(n - 1, dp) + self.nthfib(n - 2, dp)
            return dp[n]

    def fib(self, n: int) -> int:
        if n == 0:
            return 0
        dp = [-1] * (n + 1) 
        dp[0], dp[1] = 0, 1   
        
        return self.nthfib(n, dp)

File: 509-Fibonacci-Number.cpp
class Solution {
public:
    int fib(int n) {
        int first = 0, second = 1; int ans = 0;
        if( n==0 || n==1 ) return n;
        for(int i = 1; i<n; i++){
            ans = first + second;
            first = second;
            second = ans;
        }
        return ans;
    }
};